frontend/test/pytest/test_autograph.py

```diff
@@ -1343,6 +1343,153 @@ def f1():
         with pytest.raises(RuntimeError):
             qjit(autograph=True)(f1)()
 
+    def test_uninitialized_variables(self, monkeypatch):
+        """Verify errors for (potentially) uninitialized loop variables."""
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+
+        def f(pred: bool):
+            while pred:
+                x = 3
+
+            return x
+
+        with pytest.raises(AutoGraphError, match="'x' is potentially uninitialized"):
+            qjit(autograph=True)(f)
+
+    def test_init_with_invalid_jax_type(self, monkeypatch):
+        """Test loop carried values initialized with an invalid JAX type."""
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+
+        def f(pred: bool):
+            x = ""
+
+            while pred:
+                x = 3
+
+            return x
+
+        with pytest.raises(AutoGraphError, match="'x' was initialized with type <class 'str'>"):
+            qjit(autograph=True)(f)
+
+    def test_init_with_mismatched_type(self, monkeypatch):
+        """Test loop carried values initialized with a mismatched type compared to the values used
+        inside the loop."""
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+
+        def f(pred: bool):
+            x = 0.0
+
+            while pred:
+                x = 3
+
+            return x
+
+        with pytest.raises(AutoGraphError, match="'x' was initialized with the wrong type"):
+            qjit(autograph=True)(f)
+
+
+@pytest.mark.tf
+class TestFallback:
+    """Test that Python fallbacks still produce correct results."""
+
+    def test_postbinding_errors_for(self):
```

Comment:
@dime10 Could you please check the quantum-tracing test? I think it fails for me, even with your fix. I guest that the reason for this is the late binding of jax primitives in this mode.

``` python
    def test_postbinding_errors_for(self):

        with pytest.warns(
            UserWarning, match="Tracing of an AutoGraph converted for loop failed with an exception"
        ):
            @qjit(autograph=True)
            @qml.qnode(qml.device("lightning.qubit", wires=2))
            def f_jit():
                arr = jnp.array([1.0, 2.0])
                for _ in range(2):
                    # fails result verification, will trigger fallback
                    # would raise an error during lowering if left in the JAXPR
                    arr = jnp.kron(arr, arr)
                    qml.RX(jnp.mean(arr), wires=0)
                # return qml.probs()
                return arr

            arr = jnp.array([1, 2])
            expected = jnp.kron(*([jnp.kron(arr, arr)] * 2))
            result = f_jit()
            assert np.allclose(result, expected)
            print(f_jit.mlir)
```

Reply:
This should work now :)